src/core/socket.c

```diff
@@ -141,11 +144,18 @@ void socket_free_ports(Socket *s) {
 
 static void socket_done(Unit *u) {
         Socket *s = SOCKET(u);
+        SocketPeer *p;
+        Iterator i;
 
         assert(s);
 
         socket_free_ports(s);
 
+        HASHMAP_FOREACH(p, s->peers_by_address, i)
+                hashmap_remove(s->peers_by_address, p);
```

Comment:
It's important to somehow encode in the SocketPeer object whether it is currently included in a Socket unit's "peers_by_address" hashmap or not. A natural way to encode this would be by using the "socket" ptr that is already part of SocketPeer. Simply let it point ot the right Socket* structure as long as it's Socket is still around. Reset it to NULL as soon as its Socket is gone. Or in other words, rewrite the loop above to something like this:
```c
while (p = hashmap_steal_first(i)) 
        p->socket = NULL;
```